### ISSUE TITLE
Performance metrics of the classifier

### ISSUE THREAD

**Issue by chenty58** [google-research/bert#800]
I used run_classifier.py to perform a sentence classification task, which is similar to Cola.

My question is how to get the recall score, precision score and f1 score, since the eval file only shows the result of accuracy.

Thanks!

**Comment by sarnikowski** [comment]
You can achieve this by modifying a bit of code in "run_classifer.py". Specifically it is line (684-692):
```python
def metric_fn(per_example_loss, label_ids, logits, is_real_example):
    predictions = tf.argmax(logits, axis=-1, output_type=tf.int32)
    accuracy = tf.metrics.accuracy(
        labels=label_ids, predictions=predictions, weights=is_real_example)
    loss = tf.metrics.mean(values=per_example_loss, weights=is_real_example)
    return {
        "eval_accuracy": accuracy,
        "eval_loss": loss,
    }
```
Here you can add/replace any other TF metric and just add them to the returned dictionary.

**Comment by chenty58** [comment]
Thanks, I have add some code and successfully achieved the recall, precision and f1

**Comment by linWujl** [comment]
@chenty58 hi, how do you compute the f1 measure, thanks.

**Comment by guotong1988** [comment]
Same question.

**Comment by linWujl** [comment]
I had some try and finally got the method that seems to work, the code is bellow
`precision, update_op_precision = tf.metrics.precision(label_id_, logits)
  recall, update_op_recall = tf.metrics.recall(label_id_, logits)
  f1 = 2.*precision*recall/(precision+recall)
  f1_op = 2.*update_op_precision*update_op_recall/(update_op_precision+update_op_recall)
  eval_dict[label_list[j]] = (f1, f1_op)`

If there is any mistake, please let me know.

**Comment by Mahhos** [comment]
> Thanks, I have add some code and successfully achieved the recall, precision and f1

Can you please share your code to calculate precision, recall, and F-1 score? I have tried some different codes, I could not successfully get the scores though. 
Thank you.

**Comment by jz48** [comment]
Here is the workable code for me after testing, hopefully it could help you:

`      def metric_fn(per_example_loss, label_ids, logits, is_real_example):
        # tf.logging.info('inner res = %s', str([per_example_loss, label_ids, logits, is_real_example]))
        predictions = tf.argmax(logits, axis=-1, output_type=tf.int32)
        # tf.logging.info('predictions = %s', predictions)
        accuracy = tf.metrics.accuracy(
            labels=label_ids, predictions=predictions, weights=is_real_example)
        loss = tf.metrics.mean(values=per_example_loss, weights=is_real_example)

        precision, update_op_precision = tf.metrics.precision(labels=label_ids,
                                                              predictions=predictions, weights=is_real_example)

        recall, update_op_recall = tf.metrics.recall(labels=label_ids,
                                                     predictions=predictions, weights=is_real_example)
        f1 = 2. * precision * recall / (precision + recall)
        f1_op = 2. * update_op_precision * update_op_recall / (update_op_precision + update_op_recall)

        return {
            "eval_accuracy": accuracy,
            "eval_loss": loss,
            "precision": (precision, update_op_precision),
            "recall": (recall, update_op_recall),
            "f1 score": (f1, f1_op),
        }
`
